File: starters/CSharp/Bomberjam/Game.cs
using System;
using System.Globalization;
using System.Text.Json;

namespace MyBot.Bomberjam
{
    public class Game
    {
        private bool _isReady;
        private string _myPlayerId;
        private State _state;

        public Game()
        {
            this._isReady = false;
            this._myPlayerId = null;
            this._state = null;
        }

        public string MyPlayerId
        {
            get
            {
                this.EnsureIsReady();
                return this._myPlayerId;
            }
        }

        public State State
        {
            get
            {
                this.EnsureIsReady();
                this.EnsureInitialState();
                return this._state;
            }
        }

        public Player MyPlayer
        {
            get
            {
                this.EnsureIsReady();
                this.EnsureInitialState();
                return this._state.Players[this._myPlayerId];
            }
        }

        public void Ready(string playerName)
        {
            if (this._isReady)
                return;

            if (string.IsNullOrWhiteSpace(playerName))
                throw new ArgumentException("Your name cannot be null or empty");

            Console.Out.WriteLine("0:{0}", playerName);
            Console.Out.Flush();

            this._myPlayerId = Console.In.ReadLine();

            int parsedId;
            if (string.IsNullOrWhiteSpace(this._myPlayerId) || !int.TryParse(this._myPlayerId, NumberStyles.Integer, CultureInfo.InvariantCulture, out parsedId))
                throw new InvalidOperationException("Could not retrieve your ID from standard input");

            this._isReady = true;
        }

        public void ReceiveCurrentState()
        {
            this.EnsureIsReady();
            this._state = JsonSerializer.Deserialize<State>(Console.In.ReadLine());
        }

        public void SendAction(ActionKind action)
        {
            this.EnsureIsReady();
            this.EnsureInitialState();

            var actionStr = Constants.ActionKindToActionStringMappings[action];

            Console.Out.WriteLine("{0}:{1}", this._state.Tick.ToString(CultureInfo.InvariantCulture), actionStr);
            Console.Out.Flush();
        }

        private void EnsureIsReady()
        {
            if (!this._isReady)
                throw new InvalidOperationException("You need to call Game.Ready(...) with your name first");
        }

        private void EnsureInitialState()
        {
            if (this._state == null)
                throw new InvalidOperationException("You need to call Game.ReceiveCurrentState() to retrieve the initial state of the game");
        }
    }
}

File: starters/Python3/bomberjam/game.py
import json

from bomberjam.player import Player
from bomberjam.state import State


class Game:
    def __init__(self):
        self._is_ready = False
        self._my_player_id = None
        self._state = None

    @property
    def my_player_id(self) -> str:
        self.ensure_is_ready()
        return self._my_player_id

    @property
    def state(self) -> State:
        self.ensure_is_ready()
        self.ensure_initial_state()
        return self._state

    @property
    def my_player(self) -> Player:
        self.ensure_is_ready()
        self.ensure_initial_state()
        return self._state.players[self._my_player_id]

    def ready(self, player_name):
        if self._is_ready:
            return

        if player_name is None or len(player_name.strip()) == 0:
            raise ValueError('Your name cannot be null or empty')

        print('0:' + player_name, flush=True)

        self._my_player_id = input()
        if not self._my_player_id.isdigit():
            raise RuntimeError('Could not retrieve your ID from standard input')

        self._is_ready = True

    def receive_current_state(self):
        self.ensure_is_ready()
        json_state = json.loads(input())
        self._state = State(json_state)

    def send_action(self, action):
        self.ensure_is_ready()
        self.ensure_initial_state()
        print(str(self._state.tick) + ':' + action, flush=True)

    def ensure_is_ready(self):
        if not self._is_ready:
            raise RuntimeError('You need to call Game.ready(...) with your name first')

    def ensure_initial_state(self):
        if self._state is None:
            raise RuntimeError('You need to call Game.receiveCurrentState() to retrieve the initial state of the game')
